guides/source/configuring.md

```diff
@@ -764,6 +764,29 @@ regular expressions.
 
 Specifies if source locations of methods that call database queries should be logged below relevant queries. By default, the flag is `true` in development and `false` in all other environments.
 
+#### `config.active_record.async_query_executor`
+
+Specifies how asynchronous queries are pooled.
+
+It defaults to `nil`, which means `load_async` is disabled and instead directly executes queries in the foreground.
+For queries to actually be performed asynchronously, it must be set to either `:global_thread_pool` or `:multi_thread_pool`.
+
+`:global_thread_pool` will use a single pool for all databases the application connects to. This is the preferred configuration
+for applications with only a single database, or applications which only ever query one database shard at a time.
+
+`:multi_thread_pool` will use one pool per database, and each pool size can be configured individually in `database.yml` through the
+`max_threads` and `min_thread` properties. This can be useful to applications regularly querying multiple databases at a time, and that need to more precisely define the max concurrency.
+
+#### `config.active_record.global_executor_concurrency`
+
+Used in conjunction with `config.active_record.async_query_executor = :global_thread_pool`, defines how many asynchronous
+queries can be executed concurrently.
+
+Defaults to `4`.
+
+This number must be considered in accordance with the database pool size configured in `database.yml`. The connection pool
+should be large enough to accommodate both the foreground threads (.e.g web server or job worker threads) and background threads.
```

Comment:
@byroot - @joshuay03 and I are trying to wrap our heads around what this means. Maybe an example would be helpful.

Our assumption is that the `pool` config in `database.yml` is per-process, which is why by default it matches the max thread count: 5 in `puma.rb` (for now), 5 in `database.yml`.

From what I've read (eg. [here](https://github.com/rails/rails/issues/50450#issuecomment-1880225809)), if you're using async queries then you should probably have a pool of at least 10, so that each puma thread can also run 1 async query.

But then if this defaults to 4, then based on this logic you probably need a pool of 25.

Is this what you are trying to say here? If so could I suggest this wording:

```
This number must be considered in accordance with the database pool size configured in `database.yml`.
The connection pool should be large enough to accommodate both the foreground threads (.e.g web server or job worker threads) and background threads.
You can calculate this by doing `(thread_count + 1) * global_executor_concurrency`

For example, if your web server uses 3 threads per process, and you use the default setting (4) here, then in `database.yml` your `pool` config should be 16.
```

Reply:
> But then if this defaults to 4, then based on this logic you probably need a pool of 25.

No, it's 4 threads in addition to the 5 puma threads. The pool is global, each puma thread doesn't have its own async threads pool. They all share the same process global one.

So it's `(puma_thread_count + 1) + global_executor_concurrency`, not `*`